api/channel.go

```diff
@@ -427,237 +414,115 @@ func leave(c *Context, w http.ResponseWriter, r *http.Request) {
-			Type:      model.POST_CHANNEL_DELETED,
-			UserId:    c.Session.UserId,
-		}
-
-		if _, err := app.CreatePost(post, c.TeamId, false); err != nil {
-			l4g.Error(utils.T("api.channel.delete_channel.failed_post.error"), err)
-		}
-
-		now := model.GetMillis()
-		for _, hook := range incomingHooks {
-			if result := <-app.Srv.Store.Webhook().DeleteIncoming(hook.Id, now); result.Err != nil {
-				l4g.Error(utils.T("api.channel.delete_channel.incoming_webhook.error"), hook.Id)
-			}
-		}
-
-		for _, hook := range outgoingHooks {
-			if result := <-app.Srv.Store.Webhook().DeleteOutgoing(hook.Id, now); result.Err != nil {
-				l4g.Error(utils.T("api.channel.delete_channel.outgoing_webhook.error"), hook.Id)
-			}
-		}
-
-		if dresult := <-app.Srv.Store.Channel().Delete(channel.Id, model.GetMillis()); dresult.Err != nil {
-			c.Err = dresult.Err
-			return
-		}
-		app.InvalidateCacheForChannel(channel.Id)
-
-		c.LogAudit("name=" + channel.Name)
-
-		message := model.NewWebSocketEvent(model.WEBSOCKET_EVENT_CHANNEL_DELETED, c.TeamId, "", "", nil)
-		message.Add("channel_id", channel.Id)
+	err = app.DeleteChannel(channel, c.Session.UserId)
+	if err != nil {
+		c.Err = err
+		return
+	}
 
-		app.Publish(message)
+	c.LogAudit("name=" + channel.Name)
 
-		result := make(map[string]string)
-		result["id"] = channel.Id
-		w.Write([]byte(model.MapToJson(result)))
-	}
+	result := make(map[string]string)
+	result["id"] = channel.Id
+	w.Write([]byte(model.MapToJson(result)))
 }
 
 func getChannel(c *Context, w http.ResponseWriter, r *http.Request) {
 	params := mux.Vars(r)
 	id := params["channel_id"]
 
-	cchan := app.Srv.Store.Channel().Get(id, true)
-	cmchan := app.Srv.Store.Channel().GetMember(id, c.Session.UserId)
-
-	if cresult := <-cchan; cresult.Err != nil {
-		c.Err = cresult.Err
+	var channel *model.Channel
+	var err *model.AppError
+	if channel, err = app.GetChannel(id); err != nil {
+		c.Err = err
 		return
-	} else if cmresult := <-cmchan; cmresult.Err != nil {
-		c.Err = cmresult.Err
+	}
+
+	var member *model.ChannelMember
+	if member, err = app.GetChannelMember(id, c.Session.UserId); err != nil {
+		c.Err = err
 		return
-	} else {
-		data := &model.ChannelData{}
-		data.Channel = cresult.Data.(*model.Channel)
-		member := cmresult.Data.(model.ChannelMember)
-		data.Member = &member
+	}
 
-		if data.Channel.TeamId != c.TeamId && data.Channel.Type != model.CHANNEL_DIRECT {
-			c.Err = model.NewLocAppError("getChannel", "api.channel.get_channel.wrong_team.app_error", map[string]interface{}{"ChannelId": id, "TeamId": c.TeamId}, "")
-			return
-		}
+	data := &model.ChannelData{}
+	data.Channel = channel
+	data.Member = member
 
-		if HandleEtag(data.Etag(), "Get Channel", w, r) {
-			return
-		} else {
-			w.Header().Set(model.HEADER_ETAG_SERVER, data.Etag())
-			w.Write([]byte(data.ToJson()))
-		}
+	if data.Channel.TeamId != c.TeamId && data.Channel.Type != model.CHANNEL_DIRECT {
```

Comment:
Maybe move this up before getting the channel member as we could prevent a query call if it will return this error?